build/purger/purger.py

```diff
@@ -229,6 +274,194 @@ def wait_and_purge_room(room_id: str = None, event_id: str = None) -> None:
+                room_id = api.get_room_id(local_room_alias)
+                broadcast_room_infos.append(RoomInfo(room_id, broadcast_room_alias, server))
+            except MatrixError as ex:
+                click.secho(f"Could not find room {broadcast_room_alias} with error {ex}")
+
+    return network_to_broadcast
+
+
+def update_user_activity(
+    api: GMatrixHttpApi,
+    global_user_activity: Dict[str, Any],
+    broadcast_rooms: Dict[str, List[RoomInfo]],
+) -> Dict[str, List[str]]:
+    """
+    runs update on users' presences which are about to be kicked.
+    If they are still overdue they are going to be added to the list
+    of users to be kicked. Presence updates are included in the
+    global list which is going to be stored in the user_activity_file
+    :param api: api to its own server
+    :param global_user_activity: content of user_activity_file
+    :param broadcast_rooms: room iinformation for broadcast rooms
+    :return: a list of due users to be kicked
+    """
+    current_time = int(time.time())
+    last_user_activity_update = global_user_activity["last_update"]
+    network_to_users = global_user_activity["network_to_users"]
+    network_to_due_users = dict()
+    fetch_new_members = False
+
+    # check if new members have to be fetched
+    # new members only have to be fetched every
+    # USER_PURGING_THRESHOLD since that is the earliest time
+    # new members are able to be kicked. This keeps the load on
+    # the server as leightweight as possible
+    if last_user_activity_update < current_time - USER_PURGING_THRESHOLD:
+        fetch_new_members = True
+        global_user_activity["last_update"] = current_time
+
+    for network_key, user_activity in network_to_users.items():
+        if fetch_new_members:
+            _fetch_new_members_for_network(
+                api=api,
+                user_activity=network_to_users[network_key],
+                discovery_room=broadcast_rooms[network_key][0],
+                current_time=current_time,
+            )
+
+        network_to_due_users[network_key] = _update_user_activity_for_network(
+            api=api, user_activity=user_activity, current_time=current_time
+        )
+
+    return network_to_due_users
+
+
+def _fetch_new_members_for_network(
+    api: GMatrixHttpApi, user_activity: Dict[str, int], discovery_room: RoomInfo, current_time: int
+) -> None:
+    try:
+        response = api.get_room_members(discovery_room.room_id)
+        room_members = [
+            event["state_key"]
+            for event in response["chunk"]
+            if event["content"]["membership"] == "join"
+            and event["state_key"].split(":")[1] == urlparse(api.base_url).netloc
+        ]
+
+        # Add new members with an overdue activity time
+        # to trigger presence update later
+        for user_id in room_members:
+            if user_id not in user_activity:
+                user_activity[user_id] = current_time - USER_PURGING_THRESHOLD
```

Comment:
This seems like a quite indirect to force an update. Can't we update the activity right away or assume that the user is active now, since we see him for the first time?

Reply:
Actually it is a force for an update.

Of course the presence call could happen here as well. I intended to abstract the logic from user fetching and the presence update. In fact we do assume that the user is offline and not online. 

That is due to the fact that new members get fetched in an interval as large as possible,. Which is once per USER_PURGING_THRESHOLD. That is the earliest point in time when a user can become offline. To mark him as offline is simply done to fetch the presence in the next step where all presences for due users are fetched. This includes the new users in the normal process.